mathics/algorithm/parts.py

```diff
@@ -18,7 +18,7 @@
     PartRangeError,
 )
 
-SymbolNothing = Symbol("Nothing")
+SymbolNothing = Constant("Nothing")
```

Comment:
Here and in other places `Symbol` is fine. `Constant` is not a singleton, but `Symbol` is.

Reply:
We need to distinguish when an object can change its final value or not. In many languages like lisp, Symbol is intended to be something whose value isn't inspected. The item is tested in totality.  

In Mathematica though Symbol is often used for this other idea of something that holds an unknown value which _can_ expected to change over time. It is that property we want to separate. Whether you call it Symbol("Nothing") or Constant("Nothing") it is value doesn't change. 

In fact though there are 3 distinct kinds of objects: those that have a single name an there is no more information and this doesn't change. True and False are like that. Then there are things that have a value but it is expected it shouldn't be altered after its _initial_ value. Numbers and string are like that. And then there are these placeholders for values that can have values, usually these start out unknown and somethings the value gets resolved later, and it might even get changed to a different value later.  From the evaluation standpoint that's what we are trying to distinguish here. 

Singleton is along a different dimension and has to do with implementation efficiency - whether all uses of a value share the same location or now. Since it seems preferable that Constant should be implemented a singleton kind of object that should be done.

Reply:
@rocky, in WL, a `Symbol` is an atomic expression that can have associated rules and attributes. There are just a few of symbols (like `True`, `False` or `List`) (those which have the attribute `Locked`) which can not be assigned under any circumstance. All the other symbols can be unproctected and assigned with a value. For example
``
In[1]:=Sin[Pi]
Out[1] = 0
In[2]:=Unprotect[Pi]; Pi=4; Sin[Pi]
Out[2] = Sin[4]
``
Notice also that the attribute `Constant` does not mean that a symbol has always the same value, but that in expressions like ``Dt[F[Pi,x]]`` the derivative regarding  `Pi` always vanishes.

Reply:
Ok thanks for chekcing. But we still have things like Numbers, Integers, Reals, Complex, Strings with constant or literal values.

Reply:
@rocky three problem is: with the Constant attribute we can easily use non constant and constant symbols. If we singletonize Constant we would need to check for symbols in 4 dictionaries (we want it to be just 1).

Reply:
In WL, `Symbol`s do not have a "value" (or any other attribute, apart its name). This are atoms (like numbers and strings) that can compose an `Expression`. The specific aspect of a `Symbol` is that they have associated `Definition`s. 

 In WMA, each `Symbol` has a unique `Definition`, with its attributes and replace rules ("values"). In Mathics, a `Symbol` could have at the same time several `Definition`s (because we handle different "sessions" in the same process). 

Indeed, you can erase a `Definition` at some point, but you can not erase a `Symbol`, since it can be a part of existent `Expression`s. 

Comming back to the motivation of this PR: `Symbol`s are not `Constants` or `Variable` by itself. If a `Symbol` represents a `Constant` or not, depends on its `Definition`, which can be different in different sessions.

One way to reduce the overhead of the evaluation sequence  Expression-> lookup_name->definition->rules->apply rule  would be then to change the way in which we implement different sessions (which is important for Mathics-Django) in a similar way in which WMA works: having kernels as an independent process. Then, on each instance, each `Symbol` would have just one `Definition`, and hence, a `Symbol` can be tagged as `Constant`.

Reply:
What is becoming clear is that there are three separate concepts that Mathics uses and confuses some of these. 

`Symbol` can be Mathics' class object for a WL Symbol. But in implementing this and in _implementing_ expressions and the system in general we have the concept of a Lisp kind of Symbol which I suggest calling `SymbolName`. Think of it at as an object where the only thing you can do is ask to bring it up and see if it is the same as some other name. And there is the concept of something that holds a unchangeable literal value, let's call that `LiteralValue`. 

Literal values never need reevaluation. The value that comes out when you evaluate Integer(5) or String("Foo") or True are literal values.

SymbolNames never need reevaluation when definitions haven't changed which is most of the time. 

There should be distinct implementations of each of these and we should go over the classes so that the properties of each is distinct.